signer/signer_test.go

```diff
@@ -303,3 +310,123 @@ var rejectedFileNames = []string{
 	"control*chars",
 	"_non_alpha_start",
 }
+
+func TestMakeKeyAWSECDSA(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockCtx := mockpkcs11.NewMockPKCS11Context(ctrl)
+	defer ctrl.Finish()
+
+	// we don't actually have any use for the private key besides extracting
+	// the public key from it, but I couldn't find a way to directly construct
+	// the public key.
+	privKey, err := ParsePrivateKey([]byte(ecdsaPrivateKey))
+	if err != nil {
+		t.Fatalf("failed to parse private key: %v", err)
+	}
+	pubKey := privKey.(*ecdsa.PrivateKey).PublicKey
+
+	// annoyingly, we also need the ecdh form of the private key
+	// to find the length of it, to mock some things correctly.
+	ecdhPrivKey, err := privKey.(*ecdsa.PrivateKey).ECDH()
+	if err != nil {
+		t.Fatalf("failed to convert ecdsa private key to ecdh: %v", err)
+	}
+	ecdhPubKey := ecdhPrivKey.PublicKey()
+
+	// p256, prefix, and ecPointValue are all just intermediaries to set up
+	// pubKeyAttrs, which is a return value needed by one of the mocks.
+	p256, err := asn1.Marshal(asn1.ObjectIdentifier{1, 2, 840, 10045, 3, 1, 7})
+	if err != nil {
+		t.Fatalf("failed to marshal p256 object identifier")
+	}
+	// nasty hack because i couldn't find any other way to make unmarshalEcPoint
+	// happy.
+	prefix := make([]byte, 2)
+	prefix[0] = 0x04
+	prefix[1] = byte(len(ecdhPubKey.Bytes()))
+	ecPointValue := append(prefix, ecdhPubKey.Bytes()...)
+
+	pubKeyAttrs := []*pkcs11.Attribute{
+		&pkcs11.Attribute{
+			Type: pkcs11.CKA_ECDSA_PARAMS,
+			Value: p256,
+		},
+		&pkcs11.Attribute{
+			Type: pkcs11.CKA_EC_POINT,
+			Value: ecPointValue,
+		},
+	}
+
+	// These aren't under test, they just need to return sane values to
+	// make the code under test working. Because of this, we don't bother
+	// worrying about setting useful expectations for the arguments.
+	mockCtx.EXPECT().Initialize().Return(nil).AnyTimes()
+	mockCtx.EXPECT().GetSlotList(gomock.Any()).Return([]uint{0}, nil).AnyTimes()
+	mockCtx.EXPECT().GetTokenInfo(gomock.Any()).Return(pkcs11.TokenInfo{}, nil).AnyTimes()
+	mockCtx.EXPECT().OpenSession(gomock.Any(), gomock.Any()).Return(pkcs11.SessionHandle(0), nil).AnyTimes()
+	mockCtx.EXPECT().GenerateKeyPair(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+	mockCtx.EXPECT().GetAttributeValue(gomock.Any(), gomock.Any(), gomock.Any()).Return(pubKeyAttrs, nil).AnyTimes()
+	// these ones are called as part of Close(), not as part of our actual testing
+	mockCtx.EXPECT().CloseSession(gomock.Any()).AnyTimes()
+	mockCtx.EXPECT().CloseAllSessions(gomock.Any()).AnyTimes()
+	mockCtx.EXPECT().Finalize().AnyTimes()
+	mockCtx.EXPECT().Destroy().AnyTimes()
```

Comment:
These expectations, and all of the above that goes into them are not great. I'm open to alternatives.

Reply:
Why are all of these AnyTimes instead of exact? Should the GenerateKeyPair return the key we're creating in this test? A lot of these methods seem to take `gomock.Any()` that we have control over (e.g. OpenSession returns a specific SessionHandle that we would expect it to be sent to CloseSession, no?)

Reply:
Honestly - I made these quite liberal simply because `pkcs11` isn't under test. I can update them, it just seemed unnecessary to verify that `pkcs11` and/or `crypto11` is behaving well internally when `MakeKey` is what's under test.

Reply:
Well, you need them to verify that crypto11 is calling the right parts of pkcs11 in the right order, right?

Reply:
Otherwise, why are we mocking out pkcs11 isntead of crypto11?

Reply:
I would prefer to mock out crypto11! But AIUI that would require a deeper refactoring to wrap all of the `crypto11` uses in some interface we can then dependency inject.

I don't think there's a any harm in improving this test as things stand now though.

Reply:
I updated the mock expectations.